### ISSUE TITLE
[API Proposal]: Expose MemoryHandle's inner GCHandle and IPinnable

### ISSUE THREAD

**Issue by Tacodiva** [dotnet/runtime#93841]
### Background and motivation

Currently, the API provides `Memory<T>`, a reference to some memory on the heap. This memory could be anything, like a slice of a managed array, a region of unmanaged memory, or some other arbitrary memory scheme. When pinned, the low-level handle to this arbitrary memory is a `MemoryHandle`. It can hold onto a reference to arbitrary memory by storing two things, a `GCHandle` and an `IPinnable`. The `GCHandle` is used for situations where the memory referenced is a part of a managed object (like a section of an array), while the `IPinnable` is used for situations where the referenced memory is managed by an external mechanism (like a `MemoryManager<T>`). If no memory management is required for the reference, than neither of them have to be used (for example, a constant region of memory which is never deallocated wouldn't need either). 

Structs like `Memory<T>` are useful because they provide a high-level view onto something low-level. API authors get the power to deal with the nitty-gritty memory when they need to, while all the end user sees is the pretty structs. Currently however, `MemoryHandle` is pretty much completely opaque, making it unnecessarily difficult to reason about its contents on a low level.

The example that compelled me to make this suggestion is one where I want to store a `Memory<T>` in unmanaged memory. It is currently possible to to this by creating a `GCHandle` to the `MemoryHandle`, then storing that in unmanaged memory (along with the actual fixed pointer), disposing both the `MemoryHandle` and the `GCHandle` pointing to the memory handle when finished. However, in the most common usecase of my API, the `MemoryHandle` is a reference to a part of a managed array, meaning it itself is essentially just a pointer and a `GCHandle` which I could just store directly, but can't because the handle is not exposed. Instead, I am forced into unnecessarily boxing the `MemoryHandle` so I can get a `GCHandle` to it in a very hot code path.

### API Proposal

```diff
namespace System.Buffers;

public unsafe struct MemoryHandle // ...
{
    // ...
    
+    public GCHandle Handle { get; }

+    public IPinnable? Pinnable { get; }

    // ...
}
```

### API Usage

```csharp
public unsafe void MyAPI(Memory<byte> buffer)
{
    MemoryHandle bufferHandle = buffer.Pin();

    // Will will now turn our buffer handle into a GCHandle

    GCHandle handle;
    
    bool hasHandle = bufferHandle.Handle.IsAllocated;
    bool hasPinnable = bufferHandle.Pinnable != null;

    if (hasHandle && !hasPinnable)
    {
        // Only a GCHandle means we can just store the GCHandle
        handle = bufferHandle.Handle;
    } 
    else if (hasPinnable && !hasHandle)
    {
        // Only an IPinnable means we can store a GCHandle to the pinnable.
        handle = GCHandle.Alloc(bufferHandle.Pinnable);
    }
    else if (hasPinnable && hasHandle)
    {
        // If we have both, we can store a GCHandle to the boxed MemoryHandle
        // Note that I don't think this will even happen under normal circumstances but it could happen if 
        //  a custom `MemoryManager<T>` is created which returns a MemoryHandle with both.
        handle = GCHandle.Alloc(bufferHandle);
    }
    else
    {
        // If we have neither, we don't need to store any handles
        handle = default;
    }

    StoreUnmanaged(bufferHandle.Pointer, handle);
}
```

### Alternative Designs

For my use case, it is possible to write a new struct which represents a `MemoryHandle` but is able to be stored in unmanaged memory (that's probably what I would end up doing anyway), but I don't know that this use is common enough to warrant that, and there are different implementation approaches that one might take depending on the usecase. It is my opinion that with these low-level APIs, there's very little reason to hide things that could be used to improve performance, which is exactly what `MemoryHandle` is doing.

### Risks

(Not sure if this is a 'risk' but...) I did notice that the design I've suggested introduces what is a bit of an anti-pattern regarding `MemoryHandle.Dispose`. If you are pulling the `GCHandle` and `IPinnable` out of the `MemoryHandle`, you shouldn't dispose it because that will free the handle / unpin the reference. I'm not sure how big of a problem that is, but thought I'd include it here.

**Comment by Clockwork-Muse** [comment]
... Would it be beneficial to you to just use native memory, and then expose it to C#?  That is, invert the storage direction?

**Comment by Tacodiva** [comment]
> That is, invert the storage direction?

Yes for sure, it would be easier if i did that, but I would prefer it if the consumer of the API I'm creating didn't have to deal with that. Just passing in the managed `Memory<T>` is so much cleaner.
For some context, I'm coding a graphics API where the model data (like vertices and stuff) needs to be held by unmanaged memory while a frame is being rendered. I'd really like it if you could just pass in a `Memory<T>` and I could just use it as is without having to copy everything.

**Comment by tannergooding** [comment]
Closing this given little engagement and it being a large change in the design and implications of memory handle. I don't see the justification for driving it through. In many cases you can just cache the handle from when you create it.
